h3-quinn/src/lib.rs

```diff
@@ -1,223 +1,149 @@
+use quinn::VarInt;
+
+pub use quinn::{self};
 
-/// A QUIC connection backed by Quinn
+/// QUIC connection
 ///
-/// Implements a [`quic::Connection`] backed by a [`quinn::Connection`].
+/// A [`quic::Connection`] backed by [`quinn::Connection`].
 pub struct Connection {
     conn: quinn::Connection,
-    incoming_bi: IncomingBiStreams,
-    opening_bi: Option<OpenBi>,
-    incoming_uni: IncomingUniStreams,
-    opening_uni: Option<OpenUni>,
 }
 
 impl Connection {
-    /// Create a [`Connection`] from a [`quinn::NewConnection`]
-    pub fn new(new_conn: NewConnection) -> Self {
-        let NewConnection {
-            uni_streams,
-            bi_streams,
-            connection,
-            ..
-        } = new_conn;
-
-        Self {
-            conn: connection,
-            incoming_bi: bi_streams,
-            opening_bi: None,
-            incoming_uni: uni_streams,
-            opening_uni: None,
-        }
+    /// Create a [`Connection`] from a [`quinn::Connection`]
+    pub fn new(conn: quinn::Connection) -> Self {
+        Self { conn }
     }
 }
 
-/// The error type for [`Connection`]
-///
-/// Wraps reasons a Quinn connection might be lost.
-#[derive(Debug)]
-pub struct ConnectionError(quinn::ConnectionError);
-
-impl std::error::Error for ConnectionError {}
-
-impl fmt::Display for ConnectionError {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.0.fmt(f)
-    }
-}
-
-impl Error for ConnectionError {
-    fn is_timeout(&self) -> bool {
-        matches!(self.0, quinn::ConnectionError::TimedOut)
-    }
-
-    fn err_code(&self) -> Option<u64> {
-        match self.0 {
-            quinn::ConnectionError::ApplicationClosed(quinn_proto::ApplicationClose {
-                error_code,
-                ..
-            }) => Some(error_code.into_inner()),
-            _ => None,
-        }
-    }
-}
-
-impl From<quinn::ConnectionError> for ConnectionError {
-    fn from(e: quinn::ConnectionError) -> Self {
-        Self(e)
-    }
-}
-
-impl<B> quic::Connection<B> for Connection
-where
-    B: Buf,
-{
+impl<B: Buf> quic::Connection<B> for Connection {
+    type OpenStreams = OpenStreams;
+    type BidiStream = BidiStream<B>;
     type SendStream = SendStream<B>;
     type RecvStream = RecvStream;
-    type BidiStream = BidiStream<B>;
-    type OpenStreams = OpenStreams;
     type Error = ConnectionError;
 
     fn poll_accept_bidi(
         &mut self,
-        cx: &mut task::Context<'_>,
+        cx: &mut Context<'_>,
     ) -> Poll<Result<Option<Self::BidiStream>, Self::Error>> {
-        let (send, recv) = match ready!(self.incoming_bi.next().poll_unpin(cx)) {
-            Some(x) => x?,
-            None => return Poll::Ready(Ok(None)),
-        };
-        Poll::Ready(Ok(Some(Self::BidiStream {
-            send: Self::SendStream::new(send),
-            recv: Self::RecvStream::new(recv),
-        })))
+        Poll::Ready(
+            match ready!(Box::pin(self.conn.accept_bi()).poll_unpin(cx)) {
```

Comment:
How about pining it on the stack:

```rust
let mut future = self.conn.accept_bi();
let future = unsafe { Pin::new_unchecked(&mut future);
```

And poll the future directly:

```rust
future.poll(cx)
```

This is essentially what `pin_mut!` does.

Reply:
I think this code causes a problem, since it's creating a future (calling an async fn) on poll, polling it once, and if it wasn't ready, the future is dropped since it's in the function stack. To translate correctly to polling, the future would need to be store in self somehow.

Reply:
> I think this code causes a problem, since it's creating a future (calling an async fn) on poll, polling it once, and if it wasn't ready, the future is dropped since it's in the function stack. To translate correctly to polling, the future would need to be store in self somehow.
> 
> This problem is in each of the poll functions.

This is probably the reason why the connections timed out. Every time I tested this and the future was ready it worked. If the future was not ready it timed out.

Reply:
@seanmonstar I see. It totally makes sense. I feel I may need further guidance for this. I don't see a clear path for storing the futures in self without changing the traits as they require lifetime annotations now.

Curious, why did the tests pass on macOS?

Reply:
So, we need to somehow “park” the futures. It looks like what an async executor need to do, so I wonder if `tokio` would provide some tools for that.

Reply:
> @seanmonstar I see. It totally makes sense. I feel I may need further guidance for this. I don't see a clear path for storing the futures in self without changing the traits as they require lifetime annotations now.
> 
> Curious, why did the tests pass on macOS?

what would speak against changing the traits? The crate is not released yet, right?